### DISCUSSION TITLE
Shift reg with blackpill

### DISCUSSION THREAD

**Discussion by body32003** [grblHAL/STM32F4xx#132] [Q&A]
Hello
First of thank you for the awesome work
I'm still using grbl mega 5x
Secondly I'm looking into building a blackpill cnc board that uses 74hc595 shift reg to expand the outputs and step dir pulses 
Can it be done

**Comment by terjeio** [comment]
> Can it be done

Yes, should be doable. However, If for a one-off project then IMO it is not worth it unless it is a coding excercise.

FYI I made a patch to [output XY position to audio DACs via I2S DMA](https://github.com/grblHAL/STM32F4xx/blob/master/Src/st_morpo_dac.c) as an experiment, this without any changes to _driver.c_. I2S DMA could be used for step/dir output in a similar way as how it is done for the [ESP32](https://github.com/grblHAL/ESP32/blob/master/main/i2s_out.c), and the experimental code might help you to get started.

**Reply by body32003** [reply]
Yeah I noticed that bearing's esp 32 uses i2s while Picocnc uses spi 
I'll try that for blackpill
And who know maybe I'll be able to make a replicable thing that's easy enough for ppl to use 
Maybe a shield of some sort for the blackpill to mount a ramps 1.4

**Reply by terjeio** [reply]
> Picocnc uses spi

FYI it uses [PIO programs](https://github.com/grblHAL/RP2040/blob/e9910d32141044ea133a9e43f55c7e24e3978dea/driverPIO.pio#L110-L220) to generate and output the pulse and direction signals with programmable length and delay, not plain SPI.
